File: backend/src/services/elo.rs
use chrono::DateTime;
use sqlx::PgPool;
use std::collections::HashMap;
use uuid::Uuid;

type EloConfigRow = (String, f64, f64, Option<f64>, Option<f64>, Option<i32>);

#[derive(Debug, Clone)]
pub struct EloConfig {
    pub version_name: String,
    pub k_factor: f64,
    pub starting_elo: f64,
    pub base_k_factor: Option<f64>,
    pub new_player_k_bonus: Option<f64>,
    pub new_player_bonus_period: Option<i32>,
}

#[derive(Debug)]
struct Game {
    id: Uuid,
    player1_id: Uuid,
    player2_id: Uuid,
    played_at: DateTime<chrono::Utc>,
}

/// Calculate dynamic K-factor based on player experience
/// Formula: K = base_k + (new_player_k_bonus * e^(-games_played / bonus_period))
fn calculate_dynamic_k_factor(config: &EloConfig, games_played: i32) -> f64 {
    // If dynamic K-factor is not configured, use static k_factor
    if let (Some(base_k), Some(bonus), Some(period)) = (
        config.base_k_factor,
        config.new_player_k_bonus,
        config.new_player_bonus_period,
    ) && period > 0
    {
        let decay = (-games_played as f64 / period as f64).exp();
        return base_k + (bonus * decay);
    }

    // Fallback to static k_factor
    config.k_factor
}

/// Recalculate all ELO ratings using the specified configuration
pub async fn recalculate_all_elo(
    pool: &PgPool,
    config: &EloConfig,
    job_id: Option<Uuid>,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    tracing::info!("Starting ELO recalculation with config: {:?}", config);

    // Get all players
    let players: Vec<(Uuid,)> = sqlx::query_as("SELECT id FROM players")
        .fetch_all(pool)
        .await?;

    tracing::info!("Found {} players", players.len());

    // Initialize ELO and games played for all players
    let mut player_elos: HashMap<Uuid, f64> = HashMap::new();
    let mut player_games_played: HashMap<Uuid, i32> = HashMap::new();
    for (player_id,) in players {
        player_elos.insert(player_id, config.starting_elo);
        player_games_played.insert(player_id, 0);
    }

    // Get all games in chronological order
    let games: Vec<Game> = sqlx::query_as::<_, (Uuid, Uuid, Uuid, DateTime<chrono::Utc>)>(
        "SELECT id, player1_id, player2_id, played_at FROM games ORDER BY played_at ASC",
    )
    .fetch_all(pool)
    .await?
    .into_iter()
    .map(|(id, p1, p2, played_at)| Game {
        id,
        player1_id: p1,
        player2_id: p2,
        played_at,
    })
    .collect();

    tracing::info!("Found {} games to process", games.len());

    // Start transaction for atomic update
    let mut tx = pool.begin().await?;

    // Delete old ELO history for this version
    sqlx::query("DELETE FROM elo_history WHERE elo_version = $1")
        .bind(&config.version_name)
        .execute(&mut *tx)
        .await?;

    tracing::info!(
        "Cleared existing ELO history for version '{}'",
        config.version_name
    );

    // Process each game
    for (i, game) in games.iter().enumerate() {
        let winner_elo_before = *player_elos
            .get(&game.player1_id)
            .ok_or_else(|| format!("Player {} not found in ELO map", game.player1_id))?;
        let loser_elo_before = *player_elos
            .get(&game.player2_id)
            .ok_or_else(|| format!("Player {} not found in ELO map", game.player2_id))?;

        // Get games played for each player
        let winner_games = *player_games_played
            .get(&game.player1_id)
            .ok_or_else(|| format!("Player {} not found in games played map", game.player1_id))?;
        let loser_games = *player_games_played
            .get(&game.player2_id)
            .ok_or_else(|| format!("Player {} not found in games played map", game.player2_id))?;

        // Calculate dynamic K-factors for both players
        let winner_k = calculate_dynamic_k_factor(config, winner_games);
        let loser_k = calculate_dynamic_k_factor(config, loser_games);

        // Calculate expected scores
        let expected_winner =
            1.0 / (1.0 + 10_f64.powf((loser_elo_before - winner_elo_before) / 400.0));
        let expected_loser = 1.0 - expected_winner;

        // Calculate ELO changes with player-specific K-factors
        let winner_change = winner_k * (1.0 - expected_winner);
        let loser_change = loser_k * (0.0 - expected_loser);

        let winner_elo_after = winner_elo_before + winner_change;
        let loser_elo_after = loser_elo_before + loser_change;

        // Update in-memory ELO and games played
        player_elos.insert(game.player1_id, winner_elo_after);
        player_elos.insert(game.player2_id, loser_elo_after);
        player_games_played.insert(game.player1_id, winner_games + 1);
        player_games_played.insert(game.player2_id, loser_games + 1);

        // Update game's ELO version
        sqlx::query("UPDATE games SET elo_version = $1 WHERE id = $2")
            .bind(&config.version_name)
            .bind(game.id)
            .execute(&mut *tx)
            .await?;

        // Insert ELO history for winner
        sqlx::query(
            "INSERT INTO elo_history (player_id, game_id, elo_before, elo_after, elo_version, created_at)
             VALUES ($1, $2, $3, $4, $5, $6)"
        )
        .bind(game.player1_id)
        .bind(game.id)
        .bind(winner_elo_before)
        .bind(winner_elo_after)
        .bind(&config.version_name)
        .bind(game.played_at)
        .execute(&mut *tx)
        .await?;

        // Insert ELO history for loser
        sqlx::query(
            "INSERT INTO elo_history (player_id, game_id, elo_before, elo_after, elo_version, created_at)
             VALUES ($1, $2, $3, $4, $5, $6)"
        )
        .bind(game.player2_id)
        .bind(game.id)
        .bind(loser_elo_before)
        .bind(loser_elo_after)
        .bind(&config.version_name)
        .bind(game.played_at)
        .execute(&mut *tx)
        .await?;

        if (i + 1) % 100 == 0 {
            tracing::info!("Processed {}/{} games", i + 1, games.len());

            // Update job progress if job_id is provided
            if let Some(jid) = job_id
                && let Err(e) = crate::services::jobs::update_job_progress(
                    pool,
                    jid,
                    (i + 1) as i32,
                    games.len() as i32,
                )
                .await
            {
                tracing::warn!("Failed to update job progress: {}", e);
            }
        }
    }

    // Update all players' current ELO
    for (player_id, elo) in player_elos.iter() {
        sqlx::query("UPDATE players SET current_elo = $1 WHERE id = $2")
            .bind(elo)
            .bind(player_id)
            .execute(&mut *tx)
            .await?;
    }

    // Commit transaction
    tx.commit().await?;

    tracing::info!("Successfully recalculated ELO for {} games", games.len());

    Ok(())
}

/// Get the active ELO configuration
#[allow(dead_code)]
pub async fn get_active_config(pool: &PgPool) -> Result<Option<EloConfig>, sqlx::Error> {
    let row: Option<EloConfigRow> = sqlx::query_as(
        "SELECT version_name, k_factor, starting_elo, base_k_factor, new_player_k_bonus, new_player_bonus_period
         FROM elo_configurations WHERE is_active = true"
    )
    .fetch_optional(pool)
    .await?;

    Ok(row.map(
        |(
            version_name,
            k_factor,
            starting_elo,
            base_k_factor,
            new_player_k_bonus,
            new_player_bonus_period,
        )| EloConfig {
            version_name,
            k_factor,
            starting_elo,
            base_k_factor,
            new_player_k_bonus,
            new_player_bonus_period,
        },
    ))
}

/// Get ELO configuration by version name
pub async fn get_config_by_version(
    pool: &PgPool,
    version: &str,
) -> Result<Option<EloConfig>, sqlx::Error> {
    let row: Option<EloConfigRow> = sqlx::query_as(
        "SELECT version_name, k_factor, starting_elo, base_k_factor, new_player_k_bonus, new_player_bonus_period
         FROM elo_configurations WHERE version_name = $1"
    )
    .bind(version)
    .fetch_optional(pool)
    .await?;

    Ok(row.map(
        |(
            version_name,
            k_factor,
            starting_elo,
            base_k_factor,
            new_player_k_bonus,
            new_player_bonus_period,
        )| EloConfig {
            version_name,
            k_factor,
            starting_elo,
            base_k_factor,
            new_player_k_bonus,
            new_player_bonus_period,
        },
    ))
}

/// Enum to represent which player won a game
#[derive(Debug, Clone, Copy, serde::Serialize, serde::Deserialize)]
pub enum GameWinner {
    Player1,
    Player2,
}

/// Represents ELO changes for a single game within a match
#[derive(Debug, Clone)]
pub struct MatchEloChange {
    pub game_id: Uuid,
    pub player1_id: Uuid,
    pub player2_id: Uuid,
    pub player1_elo_before: f64,
    pub player1_elo_after: f64,
    pub player1_elo_change: f64,
    pub player2_elo_before: f64,
    pub player2_elo_after: f64,
    pub player2_elo_change: f64,
}

/// Calculate ELO changes for all games in a match sequentially
/// Each game uses the updated ELO from the previous game
pub fn calculate_match_elo_changes(
    player1_id: Uuid,
    player2_id: Uuid,
    player1_starting_elo: f64,
    player2_starting_elo: f64,
    games: Vec<(Uuid, GameWinner)>, // (game_id, winner)
    player1_k_factor: f64,
    player2_k_factor: f64,
) -> Vec<MatchEloChange> {
    let mut current_p1_elo = player1_starting_elo;
    let mut current_p2_elo = player2_starting_elo;
    let mut changes = Vec::new();

    for (game_id, winner) in games {
        // Calculate expected scores
        let expected_p1 = 1.0 / (1.0 + 10_f64.powf((current_p2_elo - current_p1_elo) / 400.0));
        let expected_p2 = 1.0 - expected_p1;

        // Determine actual scores based on winner
        let (p1_score, p2_score) = match winner {
            GameWinner::Player1 => (1.0, 0.0),
            GameWinner::Player2 => (0.0, 1.0),
        };

        // Calculate ELO changes
        let p1_change = player1_k_factor * (p1_score - expected_p1);
        let p2_change = player2_k_factor * (p2_score - expected_p2);

        let p1_after = current_p1_elo + p1_change;
        let p2_after = current_p2_elo + p2_change;

        changes.push(MatchEloChange {
            game_id,
            player1_id,
            player2_id,
            player1_elo_before: current_p1_elo,
            player1_elo_after: p1_after,
            player1_elo_change: p1_change,
            player2_elo_before: current_p2_elo,
            player2_elo_after: p2_after,
            player2_elo_change: p2_change,
        });

        // Update current ELOs for next game
        current_p1_elo = p1_after;
        current_p2_elo = p2_after;
    }

    changes
}

File: apps/api/src/routes/elo.ts
import type { FastifyInstance } from 'fastify';
import type { DB } from '../db.js';
import { makeAuthHooks } from '../auth.js';
import { calculateMatchEloChanges, dynamicKFactor, type EloConfig } from '../elo.js';
import { invalidInput } from '../errors.js';
import { createJob, getJob, markJobFinished, markJobRunning, updateJobProgressItems } from '../jobs.js';
import { asBool, nowIso, uuid } from '../util.js';

const MAX_VERSION_NAME_LENGTH = 50;
const MIN_K_FACTOR = 1.0;
const MAX_K_FACTOR = 100.0;
const MIN_STARTING_ELO = 100.0;
const MAX_STARTING_ELO = 3000.0;
const MAX_DESCRIPTION_LENGTH = 500;

interface EloConfigBody {
  version_name?: string;
  k_factor: number;
  starting_elo: number;
  base_k_factor?: number | null;
  new_player_k_bonus?: number | null;
  new_player_bonus_period?: number | null;
  description?: string | null;
}

interface EloConfigRow {
  id: string;
  version_name: string;
  k_factor: number;
  starting_elo: number;
  base_k_factor: number | null;
  new_player_k_bonus: number | null;
  new_player_bonus_period: number | null;
  description: string | null;
  is_active: number;
  created_at: string;
  created_by: string | null;
}

const mapConfig = (row: EloConfigRow) => ({
  id: row.id,
  version_name: row.version_name,
  k_factor: row.k_factor,
  starting_elo: row.starting_elo,
  base_k_factor: row.base_k_factor,
  new_player_k_bonus: row.new_player_k_bonus,
  new_player_bonus_period: row.new_player_bonus_period,
  description: row.description,
  is_active: asBool(row.is_active),
  created_at: row.created_at,
});

function validateConfigBody(body: EloConfigBody): void {
  if (body.k_factor < MIN_K_FACTOR || body.k_factor > MAX_K_FACTOR) {
    throw invalidInput(`K-factor must be between ${MIN_K_FACTOR} and ${MAX_K_FACTOR}`);
  }
  if (body.starting_elo < MIN_STARTING_ELO || body.starting_elo > MAX_STARTING_ELO) {
    throw invalidInput(`Starting ELO must be between ${MIN_STARTING_ELO} and ${MAX_STARTING_ELO}`);
  }
  if (body.description != null && body.description.length > MAX_DESCRIPTION_LENGTH) {
    throw invalidInput(`Description must be ${MAX_DESCRIPTION_LENGTH} characters or less`);
  }
  if (body.base_k_factor != null && (body.base_k_factor < MIN_K_FACTOR || body.base_k_factor > MAX_K_FACTOR)) {
    throw invalidInput(`Base K-factor must be between ${MIN_K_FACTOR} and ${MAX_K_FACTOR}`);
  }
  if (body.new_player_k_bonus != null && (body.new_player_k_bonus < 0 || body.new_player_k_bonus > MAX_K_FACTOR)) {
    throw invalidInput(`New player K bonus must be between 0 and ${MAX_K_FACTOR}`);
  }
  if (body.new_player_bonus_period != null && body.new_player_bonus_period <= 0) {
    throw invalidInput('New player bonus period must be positive');
  }
}

/**
 * Global ELO recalculation across ALL games (chronological), tagged with one
 * configuration version. Ported from services/elo.rs::recalculate_all_elo.
 */
function recalculateAllElo(db: DB, config: EloConfig, jobId: string): void {
  const players = db.prepare('SELECT id FROM players').all() as Array<{ id: string }>;
  const playerElos = new Map<string, number>();
  const gamesPlayed = new Map<string, number>();
  for (const { id } of players) {
    playerElos.set(id, config.starting_elo);
    gamesPlayed.set(id, 0);
  }

  const games = db
    .prepare('SELECT id, player1_id, player2_id, season_id, played_at FROM games ORDER BY played_at ASC')
    .all() as Array<{ id: string; player1_id: string; player2_id: string; season_id: string; played_at: string }>;

  const insertHistory = db.prepare(
    `INSERT INTO elo_history (id, player_id, game_id, elo_before, elo_after, elo_version, season_id, created_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?)`
  );
  const updateGameVersion = db.prepare('UPDATE games SET elo_version = ? WHERE id = ?');

  db.transaction(() => {
    db.prepare('DELETE FROM elo_history WHERE elo_version = ?').run(config.version_name);

    games.forEach((game, i) => {
      // game.player1_id is the winner of the game.
      const winnerBefore = playerElos.get(game.player1_id)!;
      const loserBefore = playerElos.get(game.player2_id)!;
      const winnerK = dynamicKFactor(
        config.k_factor,
        config.base_k_factor,
        config.new_player_k_bonus,
        config.new_player_bonus_period,
        gamesPlayed.get(game.player1_id) ?? 0
      );
      const loserK = dynamicKFactor(
        config.k_factor,
        config.base_k_factor,
        config.new_player_k_bonus,
        config.new_player_bonus_period,
        gamesPlayed.get(game.player2_id) ?? 0
      );

      const [change] = calculateMatchEloChanges(
        winnerBefore,
        loserBefore,
        [{ gameId: game.id, winner: 'Player1' }],
        winnerK,
        loserK
      );

      playerElos.set(game.player1_id, change.player1_elo_after);
      playerElos.set(game.player2_id, change.player2_elo_after);
      gamesPlayed.set(game.player1_id, (gamesPlayed.get(game.player1_id) ?? 0) + 1);
      gamesPlayed.set(game.player2_id, (gamesPlayed.get(game.player2_id) ?? 0) + 1);

      updateGameVersion.run(config.version_name, game.id);
      insertHistory.run(
        uuid(),
        game.player1_id,
        game.id,
        change.player1_elo_before,
        change.player1_elo_after,
        config.version_name,
        game.season_id,
        game.played_at
      );
      insertHistory.run(
        uuid(),
        game.player2_id,
        game.id,
        change.player2_elo_before,
        change.player2_elo_after,
        config.version_name,
        game.season_id,
        game.played_at
      );

      if ((i + 1) % 100 === 0) {
        updateJobProgressItems(db, jobId, i + 1, games.length);
      }
    });

    const updateElo = db.prepare('UPDATE players SET current_elo = ? WHERE id = ?');
    for (const [playerId, elo] of playerElos) {
      updateElo.run(elo, playerId);
    }
  })();
}

export function registerEloRoutes(app: FastifyInstance, db: DB): void {
  const { requireAdmin } = makeAuthHooks(db);

  const getByVersion = (versionName: string): EloConfigRow | undefined =>
    db.prepare('SELECT * FROM elo_configurations WHERE version_name = ?').get(versionName) as
      | EloConfigRow
      | undefined;

  app.post<{ Body: EloConfigBody }>(
    '/api/admin/elo-configurations',
    { preHandler: requireAdmin },
    async (request) => {
      const body = request.body;
      if (!body.version_name || body.version_name.length > MAX_VERSION_NAME_LENGTH) {
        throw invalidInput(`Version name must be 1-${MAX_VERSION_NAME_LENGTH} characters`);
      }
      validateConfigBody(body);
      if (getByVersion(body.version_name)) throw invalidInput('Version name already exists');

      const id = uuid();
      db.prepare(
        `INSERT INTO elo_configurations
           (id, version_name, k_factor, starting_elo, base_k_factor, new_player_k_bonus,
            new_player_bonus_period, description, is_active, created_at, created_by)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?)`
      ).run(
        id,
        body.version_name,
        body.k_factor,
        body.starting_elo,
        body.base_k_factor ?? null,
        body.new_player_k_bonus ?? null,
        body.new_player_bonus_period ?? null,
        body.description ?? null,
        nowIso(),
        request.user.id
      );
      return mapConfig(getByVersion(body.version_name)!);
    }
  );

  app.get('/api/admin/elo-configurations', { preHandler: requireAdmin }, async () => {
    const rows = db
      .prepare('SELECT * FROM elo_configurations ORDER BY created_at DESC')
      .all() as EloConfigRow[];
    return rows.map(mapConfig);
  });

  app.put<{ Params: { versionName: string }; Body: EloConfigBody }>(
    '/api/admin/elo-configurations/:versionName',
    { preHandler: requireAdmin },
    async (request) => {
      const body = request.body;
      validateConfigBody(body);

      const existing = getByVersion(request.params.versionName);
      if (!existing) throw invalidInput('Configuration not found');
      if (asBool(existing.is_active)) {
        throw invalidInput('Cannot update active configuration. Deactivate it first.');
      }

      db.prepare(
        `UPDATE elo_configurations
         SET k_factor = ?, starting_elo = ?, base_k_factor = ?, new_player_k_bonus = ?,
             new_player_bonus_period = ?, description = ?
         WHERE version_name = ?`
      ).run(
        body.k_factor,
        body.starting_elo,
        body.base_k_factor ?? null,
        body.new_player_k_bonus ?? null,
        body.new_player_bonus_period ?? null,
        body.description ?? null,
        request.params.versionName
      );
      return mapConfig(getByVersion(request.params.versionName)!);
    }
  );

  app.delete<{ Params: { versionName: string } }>(
    '/api/admin/elo-configurations/:versionName',
    { preHandler: requireAdmin },
    async (request) => {
      const existing = getByVersion(request.params.versionName);
      if (existing && asBool(existing.is_active)) {
        throw invalidInput('Cannot delete active configuration. Deactivate it first.');
      }
      const result = db
        .prepare('DELETE FROM elo_configurations WHERE version_name = ?')
        .run(request.params.versionName);
      if (result.changes === 0) throw invalidInput('Configuration not found');
      return { message: `Configuration '${request.params.versionName}' deleted` };
    }
  );

  app.post<{ Params: { versionName: string } }>(
    '/api/admin/elo-configurations/:versionName/activate',
    { preHandler: requireAdmin },
    async (request) => {
      if (!getByVersion(request.params.versionName)) throw invalidInput('Configuration not found');
      db.transaction(() => {
        db.prepare('UPDATE elo_configurations SET is_active = 0').run();
        db.prepare('UPDATE elo_configurations SET is_active = 1 WHERE version_name = ?').run(
          request.params.versionName
        );
      })();
      return { message: `Configuration '${request.params.versionName}' activated` };
    }
  );

  app.post<{ Params: { versionName: string } }>(
    '/api/admin/elo-configurations/:versionName/recalculate',
    { preHandler: requireAdmin },
    async (request) => {
      const row = getByVersion(request.params.versionName);
      if (!row) throw invalidInput('Configuration not found');

      const config: EloConfig = {
        version_name: row.version_name,
        k_factor: row.k_factor,
        starting_elo: row.starting_elo,
        base_k_factor: row.base_k_factor,
        new_player_k_bonus: row.new_player_k_bonus,
        new_player_bonus_period: row.new_player_bonus_period,
      };
      const jobId = createJob(db, 'elo_recalculation', request.user.id);

      setImmediate(() => {
        try {
          markJobRunning(db, jobId);
          recalculateAllElo(db, config, jobId);
          markJobFinished(db, jobId, 'completed', {
            version: config.version_name,
            message: 'Recalculation completed successfully',
          });
        } catch (err) {
          app.log.error(err, 'ELO recalculation failed');
          markJobFinished(db, jobId, 'failed', { error: `Recalculation failed: ${(err as Error).message}` });
        }
      });

      return {
        message: `Started ELO recalculation for version '${request.params.versionName}'`,
        job_id: jobId,
        version: request.params.versionName,
      };
    }
  );

  app.get<{ Params: { jobId: string } }>(
    '/api/admin/jobs/:jobId',
    { preHandler: requireAdmin },
    async (request) => {
      const job = getJob(db, request.params.jobId);
      if (!job) throw invalidInput('Job not found');
      return job;
    }
  );
}
